### PR TITLE
feat(TBD-10671): added support for generic arguments

### PR DESCRIPTION
**Pull request by hliashchuk**
**Please check if the PR fulfills these requirements**

- [ ] The commit(s) message(s) follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md#commit-message-format) ?
- [ ] Unit tests for the Java changes have been added (for bug fixes / features) ?
- [ ] TUJ for the JavaJet changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?
- [ ] The new code does not introduce new technical issues

**What is the current behavior?** (You can also link to an open issue here)

**What is the new behavior?**

**BREAKING CHANGE**

If this PR contains a breaking change, please describe the impact and migration
path for existing applications.
If not please remove this section.

**Other information**:

### CODE REVIEW

File: main/plugins/org.talend.designer.components.bigdata/components/tSqoopExport/tSqoopExport_commandline.javajet [hunk 1]
```diff
@@ -42,7 +42,19 @@
...
-	boolean dieOnError = "true".equals(ElementParameterParser.getValue(node, "__DIE_ON_ERROR__"));
+	boolean dieOnError = "true".equals(ElementParameterParser.getValue(node, "__DIE_ON_ERROR__"));
+	
+	List<Map<String,String>> genericArgs = new java.util.ArrayList<Map<String,String>>();
+	List<Map<String,String>> toolSpecificArgs = new java.util.ArrayList<Map<String,String>>();
+	if(additionalList != null){ 
+		for(Map<String,String> additionalMap : additionalList){ 
+			if (additionalMap.get("ADDITIONAL_ARGUMENT").startsWith("\"--") || additionalMap.get("ADDITIONAL_ARGUMENT").startsWith("--")){
+				toolSpecificArgs.add(additionalMap);
+			} else if (additionalMap.get("ADDITIONAL_ARGUMENT").startsWith("-") || additionalMap.get("ADDITIONAL_ARGUMENT").startsWith("\"-")) {
```
Comment (amedvedenko) [inline main/plugins/org.talend.designer.components.bigdata/components/tSqoopExport/tSqoopExport_commandline.javajet:53]: duplicate part of condition, please think about extracting variable and using utility methods like `escapeQuotes` so it allows t oavoid || part of condtition.

File: main/plugins/org.talend.designer.components.bigdata/components/tSqoopExport/tSqoopExport_commandline.javajet [hunk 2]
```diff
@@ -42,7 +42,19 @@
 	boolean overrideJavaMapping = "true".equals(ElementParameterParser.getValue(node, "__DEFINE_JAVA_MAPPING__"));
 	boolean passwordStoredInFile = "true".equals(ElementParameterParser.getValue(node, "__PASSWORD_STORED_IN_FILE__"));
 
-	boolean dieOnError = "true".equals(ElementParameterParser.getValue(node, "__DIE_ON_ERROR__"));
+	boolean dieOnError = "true".equals(ElementParameterParser.getValue(node, "__DIE_ON_ERROR__"));
+	
+	List<Map<String,String>> genericArgs = new java.util.ArrayList<Map<String,String>>();
+	List<Map<String,String>> toolSpecificArgs = new java.util.ArrayList<Map<String,String>>();
```
Comment (amedvedenko) [inline main/plugins/org.talend.designer.components.bigdata/components/tSqoopExport/tSqoopExport_commandline.javajet:48]: List and Map must also be imported but it isn't, so, maybe ArrayList also can be imported? 
Also please add whitespaces to required places.

File: main/plugins/org.talend.designer.components.bigdata/components/tSqoopExport/tSqoopExport_commandline.javajet [hunk 3]
```diff
@@ -58,19 +70,21 @@ String passwordFieldName = "__PASSWORD__";
 
 String[] cmd_<%=cid %> = new String[] {"sqoop", "export"
 <%
-boolean isNull=false; 
+for(Map<String,String> genericArg : genericArgs){
+	if(genericArg.get("ADDITIONAL_VALUE")!=null || !"".equals(genericArg.get("ADDITIONAL_VALUE"))) {
+	String arg = genericArg.get("ADDITIONAL_ARGUMENT").replaceAll("\"","") + "=" + genericArg.get("ADDITIONAL_VALUE").replaceAll("\"","");
```
Comment (amedvedenko) [inline main/plugins/org.talend.designer.components.bigdata/components/tSqoopExport/tSqoopExport_commandline.javajet:75]: please don't forget about whitespaces

File: main/plugins/org.talend.designer.components.bigdata/components/tSqoopExport/tSqoopExport_commandline.javajet [hunk 4]
```diff
@@ -58,19 +70,21 @@ String passwordFieldName = "__PASSWORD__";
...
+}
 if(isUseSpeedParallel && specificParams!=null){
-	boolean isFirstParam=true;
 	for(Map<String,String> specificArg : specificParams){ 
-		isNull=false; 
-	 	if(specificArg.get("SPECIFIC_PARAM_VALUE")==null || "".equals(specificArg.get("SPECIFIC_PARAM_VALUE"))) { 
-	 		isNull=true; 
-	 	} 
-	 	if(!isNull) { 
+	 	if(specificArg.get("SPECIFIC_PARAM_VALUE")!=null || !"".equals(specificArg.get("SPECIFIC_PARAM_VALUE"))) {
```
Comment (amedvedenko) [inline main/plugins/org.talend.designer.components.bigdata/components/tSqoopExport/tSqoopExport_commandline.javajet:83]: first part of condition is useless

Comment (amedvedenko) [reply main/plugins/org.talend.designer.components.bigdata/components/tSqoopExport/tSqoopExport_commandline.javajet:83]: and whitespaces

File: main/plugins/org.talend.designer.components.bigdata/components/tSqoopExport/tSqoopExport_commandline.javajet [hunk 5]
```diff
@@ -81,10 +95,17 @@ if(isUseSpeedParallel && isUseAdditionParams){
 %>
 
 , "--connect", <%=connection%>
-<%if(!callProcStock){%>, "--table", <%=table%><%}%> , "--username", <%=username%>
-<%if(passwordStoredInFile) {%>
+<%if(!callProcStock){
+	if((!"".equals(table.replaceAll("\"","")))){
```
Comment (amedvedenko) [inline main/plugins/org.talend.designer.components.bigdata/components/tSqoopExport/tSqoopExport_commandline.javajet:99]: please replace .equals with .isEmpty

File: main/plugins/org.talend.designer.components.bigdata/components/tSqoopExport/tSqoopExport_commandline.javajet [hunk 6]
```diff
@@ -81,10 +95,17 @@ if(isUseSpeedParallel && isUseAdditionParams){
...
+%>, "--table", <%=table%>
+<%	}
+}
+if(!"".equals(username.replaceAll("\"",""))){%>
+ , "--username", <%=username%>
+<%}
+if(passwordStoredInFile) {%>
 	, "--password-file", <%=(String)ElementParameterParser.getValue(node, "__PASSWORD_FILE__")%>
-<%} else {%>
+<%} else if (!"".equals(ElementParameterParser.getValue(node, "__PASSWORD__").replaceAll("\"",""))) {%>
```
Comment (amedvedenko) [inline main/plugins/org.talend.designer.components.bigdata/components/tSqoopExport/tSqoopExport_commandline.javajet:108]: please replace .equals with .isEmpty

File: main/plugins/org.talend.designer.components.bigdata/components/tSqoopExport/tSqoopExport_commandline.javajet (hunk 5 above)
Comment (amedvedenko) [reply main/plugins/org.talend.designer.components.bigdata/components/tSqoopExport/tSqoopExport_commandline.javajet:99]: whitespaces

File: main/plugins/org.talend.designer.components.bigdata/components/tSqoopExport/tSqoopExport_commandline.javajet (hunk 6 above)
Comment (amedvedenko) [reply main/plugins/org.talend.designer.components.bigdata/components/tSqoopExport/tSqoopExport_commandline.javajet:108]: whitespaces

File: main/plugins/org.talend.designer.components.bigdata/components/tSqoopExport/tSqoopExport_commandline.javajet [hunk 7]
```diff
@@ -128,21 +149,15 @@ if(useColumnForUpdate && !callProcStock) {
...
 
-if(additionalList != null){ 
-    for(Map<String,String> additionalMap : additionalList){ 
-        isNull=false; 
-        if(additionalMap.get("ADDITIONAL_VALUE")==null || "".equals(additionalMap.get("ADDITIONAL_VALUE"))) { 
-            isNull=true; 
-        } 
-        if(!isNull) { 
+for(Map<String,String> toolSpecificArg : toolSpecificArgs){ 
+    if(toolSpecificArg.get("ADDITIONAL_VALUE")!=null || !"".equals(toolSpecificArg.get("ADDITIONAL_VALUE"))) {
```
Comment (amedvedenko) [inline main/plugins/org.talend.designer.components.bigdata/components/tSqoopExport/tSqoopExport_commandline.javajet:153]: first part of condition is useless

File: main/plugins/org.talend.designer.components.bigdata/components/tSqoopImport/tSqoopImport_java.xml [hunk 8]
```diff
@@ -441,66 +441,13 @@
...
-						<ITEM NAME="--input-optionally-enclosed-by" VALUE="--input-optionally-enclosed-by" />
-						<ITEM NAME="--lines-terminated-by" VALUE="--lines-terminated-by" />
-						<ITEM NAME="--jar-file" VALUE="--jar-file" />
-						<ITEM NAME="--null-non-string" VALUE="--null-non-string" />
-						<ITEM NAME="--null-string" VALUE="--null-string" />
-						<ITEM NAME="--optionally-enclosed-by" VALUE="--optionally-enclosed-by" />
-						<ITEM NAME="--outdir" VALUE="--outdir" />
-						<ITEM NAME="--package-name" VALUE="--package-name" />
-						<ITEM NAME="--split-by" VALUE="--split-by" SHOW_IF="USE_SPLIT=='false'" />
-						<ITEM NAME="--warehouse-dir" VALUE="--warehouse-dir" />
```
Comment (amedvedenko) [inline main/plugins/org.talend.designer.components.bigdata/components/tSqoopImport/tSqoopImport_java.xml:492]: I see a lot of items are removed, will they be added somewhere?

File: main/plugins/org.talend.designer.components.bigdata/components/tSqoopExport/tSqoopExport_commandline.javajet (hunk 1 above)
Comment (amedvedenko) [reply main/plugins/org.talend.designer.components.bigdata/components/tSqoopExport/tSqoopExport_commandline.javajet:53]: ok, escape method can be replaced with some inlined code or all quotes can be replaced if they can't be a part of ADDITIONAL_ARGUMENT

File: main/plugins/org.talend.designer.components.bigdata/components/tSqoopExport/tSqoopExport_commandline.javajet [hunk 9]
```diff
@@ -42,7 +42,19 @@
...
-	boolean dieOnError = "true".equals(ElementParameterParser.getValue(node, "__DIE_ON_ERROR__"));
+	boolean dieOnError = "true".equals(ElementParameterParser.getValue(node, "__DIE_ON_ERROR__"));
+	
+	List<Map<String,String>> genericArgs = new java.util.ArrayList<Map<String,String>>();
+	List<Map<String,String>> toolSpecificArgs = new java.util.ArrayList<Map<String,String>>();
+	if (additionalList != null) { 
+		for (Map<String,String> additionalMap : additionalList) { 
+			if (additionalMap.get("ADDITIONAL_ARGUMENT").replaceAll("\"","").startsWith("--")) {
+				toolSpecificArgs.add(additionalMap);
+			} else if (additionalMap.get("ADDITIONAL_ARGUMENT").replaceAll("\"","").startsWith("-")) {
```
Comment (amedvedenko) [inline main/plugins/org.talend.designer.components.bigdata/components/tSqoopExport/tSqoopExport_commandline.javajet:53]: additionalMap.get("ADDITIONAL_ARGUMENT").replaceAll("\"","") this part can be replaced with variable.

File: main/plugins/org.talend.designer.components.bigdata/components/tSqoopExport/tSqoopExport_commandline.javajet (hunk 2 above)
Comment (amedvedenko) [reply main/plugins/org.talend.designer.components.bigdata/components/tSqoopExport/tSqoopExport_commandline.javajet:48]: it's still actual

File: main/plugins/org.talend.designer.components.bigdata/components/tSqoopExport/tSqoopExport_commandline.javajet [hunk 10]
```diff
@@ -58,15 +70,17 @@ String passwordFieldName = "__PASSWORD__";
...
-	 	if(!isNull) { 
+for (Map<String,String> genericArg : genericArgs) {
+	if (!"".equals(genericArg.get("ADDITIONAL_VALUE"))) {
+	String arg = genericArg.get("ADDITIONAL_ARGUMENT").replaceAll("\"","") + "=" + genericArg.get("ADDITIONAL_VALUE").replaceAll("\"","");
+%>
+		,"<%=arg%>"
+<%
+	}
+}
+if (isUseSpeedParallel && specificParams!=null) {
```
Comment (amedvedenko) [inline main/plugins/org.talend.designer.components.bigdata/components/tSqoopExport/tSqoopExport_commandline.javajet:81]: whitespaces should be around != too

File: main/plugins/org.talend.designer.components.bigdata/components/tSqoopImport/tSqoopImport_java.xml (hunk 8 above)
Comment (amedvedenko) [reply main/plugins/org.talend.designer.components.bigdata/components/tSqoopImport/tSqoopImport_java.xml:492]: please answer to this)

File: main/plugins/org.talend.designer.components.bigdata/components/tSqoopImport/tSqoopImport_java.xml [hunk 11]
```diff
@@ -441,66 +441,13 @@
...
-					</ITEMS>
-				</ITEM>
+				<ITEM NAME="ADDITIONAL_ARGUMENT" FIELD="TEXT" />
```
Comment (sponomarova) [inline main/plugins/org.talend.designer.components.bigdata/components/tSqoopImport/tSqoopImport_java.xml:446]: you replace "closed list" of arguments with a text field. What is the idea here?
To unify with "java api" part?
To extend list of supported options?

File: main/plugins/org.talend.designer.components.bigdata/components/tSqoopImport/tSqoopImport_commandline.javajet [hunk 12]
```diff
@@ -103,22 +116,40 @@ if(isUseSpeedParallel && isUseAdditionParams){
...
-, "--connect", <%=connection%>, "--username", <%=username%>
-<%if(passwordStoredInFile) {%>
+, "--connect", <%=connection%>
+<%if (!username.replaceAll("\"","").isEmpty()){%>
+	, "--username", <%=username%>
+<%}
+if(passwordStoredInFile) {%>
 	, "--password-file", <%=(String)ElementParameterParser.getValue(node, "__PASSWORD_FILE__")%>
-<%} else {%>
+<%} else if (!ElementParameterParser.getValue(node, "__PASSWORD__").replaceAll("\"","").isEmpty()) {%>
```
Comment (sponomarova) [inline main/plugins/org.talend.designer.components.bigdata/components/tSqoopImport/tSqoopImport_commandline.javajet:125]: This condition looks a little bit confusing. Could you explain it?

Comment (hliashchuk) [reply main/plugins/org.talend.designer.components.bigdata/components/tSqoopImport/tSqoopImport_commandline.javajet:125]: if password has some value other than ""

File: main/plugins/org.talend.designer.components.bigdata/components/tSqoopImport/tSqoopImport_java.xml (hunk 8 above)
Comment (hliashchuk) [reply main/plugins/org.talend.designer.components.bigdata/components/tSqoopImport/tSqoopImport_java.xml:492]: No, it was hardcoded list of sqoop oprions. Now it is open list and user enter values themselves

File: main/plugins/org.talend.designer.components.bigdata/components/tSqoopImport/tSqoopImport_java.xml (hunk 11 above)
Comment (hliashchuk) [reply main/plugins/org.talend.designer.components.bigdata/components/tSqoopImport/tSqoopImport_java.xml:446]: It's replaced with a list but with no predefined values. The idea is to extend supported options

File: main/plugins/org.talend.designer.components.bigdata/components/tSqoopExport/tSqoopExport_commandline.javajet (hunk 2 above)
Comment (sponomarova) [reply main/plugins/org.talend.designer.components.bigdata/components/tSqoopExport/tSqoopExport_commandline.javajet:48]: @hliashchuk , the same remark as Andrii did.

Comment (amedvedenko) [reply main/plugins/org.talend.designer.components.bigdata/components/tSqoopExport/tSqoopExport_commandline.javajet:48]: FYI Hanna told me that for some reason adding ArrayList to import zone doesn't work so I agree with this part

**Review by AlixMetivier — APPROVED** [review]

**Review by Redwene — APPROVED** [review]
lgtm

**Pull request merged**